Restore may write nulls to project.assets.json

### NuGet Product Used

MSBuild.exe

### Product Version

9.0.100-preview.5.24307.3

### Worked before?

9.0.100-preview.3.24204.13

### Impact

I'm unable to use this version

### Repro Steps & Context

SDK task ResolvePackageAssets throws NRE while enumerating `NuGet.ProjectModel.ProjectFileDependencyGroup.Dependencies`, which returns null. 

I'm guessing that's because the `project.assets.json` contains nulls in `projectFileDependencyGroups`, which it probably shouldn't. Somehow nuget restore managed to produce these:

```
"libraries": {},
  "projectFileDependencyGroups": {
    ".NETFramework,Version=v4.7.2": [
      null,
      null,
      null,
      null,
      null,
      null,
      "Microsoft.DotNet.Build.Tasks.VisualStudio >= 9.0.0-beta.24317.3",
      "Microsoft.Net.Compilers.Toolset >= 4.11.0-2.24270.4",
      "Microsoft.VisualStudioEng.MicroBuild.Plugins.SwixBuild >= 1.1.286"
    ]
  },
```

```
C:\Program Files\dotnet\sdk\9.0.100-preview.5.24307.3\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(266,5): error MSB4018: The "ResolvePackageAssets" task failed unexpectedly.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.NET.Build.Tasks.ProjectContext.<>c__DisplayClass31_0.<GetTopLevelDependencies>b__3(String projectFileDependency)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheWriter.ComputePackageExclusions()
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheWriter..ctor(ResolvePackageAssets task)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheReader.CreateReaderFromDisk(ResolvePackageAssets task, Byte[] settingsHash)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheReader..ctor(ResolvePackageAssets task)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.ReadItemGroups()
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.ExecuteCore()
   at Microsoft.NET.Build.Tasks.TaskBase.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() 
[D:\R4\src\VisualStudio\Setup\Roslyn.VisualStudio.Setup.csproj]
```



### Verbose Logs

Can share full binlog. DM me on Teams.
Searching for `null,`:
![image](https://github.com/NuGet/Home/assets/41759/1965540e-378c-4592-a304-bc9fb2ad84ab)

--- a/src/NuGet.Core/NuGet.LibraryModel/LibraryRange.cs
+++ b/src/NuGet.Core/NuGet.LibraryModel/LibraryRange.cs
@@ -73,42 +73,40 @@ namespace NuGet.LibraryModel
             return output;
         }
 
-        public string? ToLockFileDependencyGroupString()
+        public string ToLockFileDependencyGroupString()
         {
-            if (VersionRange is null)
-            {
-                return null;
-            }
-
             StringBuilder sb = StringBuilderPool.Shared.Rent(256);
 
             sb.Append(Name);
 
-            if (VersionRange.HasLowerBound)
+            if (VersionRange != null)
             {
-                if (VersionRange.IsMinInclusive)
+                if (VersionRange.HasLowerBound)
                 {
-                    sb.Append(" >= ");
-                }
-                else
-                {
-                    sb.Append(" > ");
+                    if (VersionRange.IsMinInclusive)
+                    {
+                        sb.Append(" >= ");
+                    }
+                    else
+                    {
+                        sb.Append(" > ");
+                    }
+
+                    if (VersionRange.IsFloating)
+                    {
+                        VersionRange.Float.ToString(sb);
+                    }
+                    else
+                    {
+                        sb.Append(VersionRange.MinVersion.ToNormalizedString());
+                    }
                 }
 
-                if (VersionRange.IsFloating)
+                if (VersionRange.HasUpperBound)
                 {
-                    VersionRange.Float.ToString(sb);
+                    sb.Append(VersionRange.IsMaxInclusive ? " <= " : " < ");
+                    sb.Append(VersionRange.MaxVersion.ToNormalizedString());
                 }
-                else
-                {
-                    sb.Append(VersionRange.MinVersion.ToNormalizedString());
-                }
-            }
-
-            if (VersionRange.HasUpperBound)
-            {
-                sb.Append(VersionRange.IsMaxInclusive ? " <= " : " < ");
-                sb.Append(VersionRange.MaxVersion.ToNormalizedString());
             }
 
             return StringBuilderPool.Shared.ToStringAndReturn(sb);

--- a/src/NuGet.Core/NuGet.LibraryModel/PublicAPI.Shipped.txt
+++ b/src/NuGet.Core/NuGet.LibraryModel/PublicAPI.Shipped.txt
@@ -121,7 +121,7 @@ NuGet.LibraryModel.LibraryRange.LibraryRange(string! name, NuGet.LibraryModel.Li
 NuGet.LibraryModel.LibraryRange.LibraryRange(string! name, NuGet.Versioning.VersionRange? versionRange, NuGet.LibraryModel.LibraryDependencyTarget typeConstraint) -> void
 NuGet.LibraryModel.LibraryRange.Name.get -> string!
 NuGet.LibraryModel.LibraryRange.Name.set -> void
-NuGet.LibraryModel.LibraryRange.ToLockFileDependencyGroupString() -> string?
+NuGet.LibraryModel.LibraryRange.ToLockFileDependencyGroupString() -> string!
 NuGet.LibraryModel.LibraryRange.TypeConstraint.get -> NuGet.LibraryModel.LibraryDependencyTarget
 NuGet.LibraryModel.LibraryRange.TypeConstraint.set -> void
 NuGet.LibraryModel.LibraryRange.TypeConstraintAllows(NuGet.LibraryModel.LibraryDependencyTarget flag) -> bool

--- a/test/NuGet.Core.Tests/NuGet.LibraryModel.Tests/LibraryRangeTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.LibraryModel.Tests/LibraryRangeTests.cs
@@ -17,13 +17,14 @@ namespace NuGet.LibraryModel.Tests
         [InlineData("[1.0.0 , 2.0.0]", "packageA >= 1.0.0 <= 2.0.0")]
         [InlineData("(1.0.0 , 2.0.0)", "packageA > 1.0.0 < 2.0.0")]
         [InlineData("(1.0.0 , 2.0.0]", "packageA > 1.0.0 <= 2.0.0")]
+        [InlineData(null, "")]
         public void LibraryRange_ToLockFileDependencyGroupString(string versionRange, string expected)
         {
             // Arrange
             LibraryRange range = new LibraryRange()
             {
                 Name = "packageA",
-                VersionRange = VersionRange.Parse(versionRange),
+                VersionRange = versionRange != null ? VersionRange.Parse(versionRange) : null,
                 TypeConstraint = LibraryDependencyTarget.Project
             };
 

--- a/test/NuGet.Core.Tests/NuGet.LibraryModel.Tests/LibraryRangeTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.LibraryModel.Tests/LibraryRangeTests.cs
@@ -17,7 +17,7 @@ namespace NuGet.LibraryModel.Tests
         [InlineData("[1.0.0 , 2.0.0]", "packageA >= 1.0.0 <= 2.0.0")]
         [InlineData("(1.0.0 , 2.0.0)", "packageA > 1.0.0 < 2.0.0")]
         [InlineData("(1.0.0 , 2.0.0]", "packageA > 1.0.0 <= 2.0.0")]
-        [InlineData(null, "")]
+        [InlineData(null, "packageA")]
         public void LibraryRange_ToLockFileDependencyGroupString(string versionRange, string expected)
         {
             // Arrange

